---
title: "Basic bulk T Cell Repertoire statistics"
format:
  html: 
    toc: true
    toc_depth: 3
    theme: flatly
    code-fold: true
    embed-resources: true
jupyter: python3
---

This notebook was generated via the `bulk-tcrseq` pipeline developed at Johns Hopkins for use by BTC scientists. 

# Report Setup

## Import pipeline parameters

Default inputs are overwritten at the command line in `modules/local/plot_simple.nf`
```{python, fold=false}
#| tags: [parameters]

project_dir='path/to/project_dir'
sample_table='path/to/sample_table.csv'
simple_stats_csv='path/to/simple_stats.csv'
v_family_csv='path/to/v_family.csv'
```

## Import libraries

```{python}
import os
from IPython.display import Image
import sys 
sys.path.append(project_dir + '/bin/')

from plot_utils import plot_timecourse2, plt_combined
```

# Sample level statistics

Below are plots showing basic T cell repertoire statistics. Each plot has a simple description about the statistic shown. However, biological interpretation of each plot is left to the user.

## Number of clones

```{python}
## display the num_clones.png image
Image(filename='num_clones.png')
```

**Figure 1. Number of clones across timepoints.** The number of clones in each sample is shown. A clone is defined as a T cell with a unique CDR3 amino acid sequence. The number of clones is shown on the y-axis and the timepoint is shown on the x-axis. Left: Boxplot, Right: matched paitent line plot.

## Clonality

```{python}
## display the clonality.png image
Image(filename='clonality.png')
```

**Figure 2. The clonality of samples across timepoints.** Clonality is a measure of T cell clonal expansion and reflects the degree to which the sample is dominated by 1 or more T cell clones. Clonality is calculated as: *Clonality = 1 - entropy(clone_counts) / log2(number of clones)*. Left: Boxplot, Right: matched paitent line plot.

## Simpson Index

```{python}
## display simpson_index_corrected.png image
Image(filename='simpson_index_corrected.png')
```

**Figure 3. Corrected Simpson Index.** The Simpson Index is a measure of diversity that takes into account the number of clones and the relative abundance of each clone in a sample. The corrected Simpson Index is calculated as: *Corrected Simpson Index = sum(clone_counts\*(clone_counts-1)) / (number of TCRs\*(number of TCRs-1))*. Left: Boxplot, Right: matched paitent line plot.

## Percent of productive TCRs

```{python}
## display pct_prod.png image
Image(filename='pct_prod.png')
```

**Figure 4. Percent of productive TCRs.** A productive TCR is a DNA/RNA sequence that can be translated into a protein sequence, i.e. it does not contain a premature stop codon or an out of frame rearrangement. The percent of productive TCRs is calculated as: *Percent productive TCRs = number of productive TCRs / number of TCRs*. Left: Boxplot, Right: matched paitent line plot.

## Average CDR3 Length

```{python}
## display cdr3_len.png image
Image(filename='cdr3_avg_len.png')
```

**Figure 5. Average CDR3 Length** The average length of the CDR3 region of the TCR. The CDR3 region is the most variable region of the TCR and is the region that determines antigen specificity. Left: Boxplot, Right: matched paitent line plot.

## V gene family usage

```{python}
## display v_gene_usage.png image
Image(filename='v_family_usage.png')
```

**Figure 6. V gene family usage.** The V gene family usage of the TCRs in each sample. The x-axis shows the timepoint collected for each individual, and the y-axis shows the proportion of TCRs that use each V gene family.

# Execution environment

```{python}
print(os.system('mamba env export --no-builds'))
```
